server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ChunkOutputStream.java

```diff
@@ -1,156 +1,147 @@
-       //     the question is how to communicate that.
-       //   - Please see https://projectreactor.io/docs/netty/release/reference/index.html#_sse.
-       //     It reads `The flushing strategy is "flush after every element" emitted
-       //     by the provided Publisher`.  So, we do not need to call flush individually.
-       //
-       //   In summary, we are not controlling flushing, instead Reactor Netty does.
-       //   Also, this#asyncWrite will manage the backpressure.  I assume Reactor Netty
-       //   will backpressure if it cannot flush, and that would mean an element won't be
-       //   requested so asyncWrite won't complete until the element is pulled.  Also,
-       //   asyncFlush is called right after an asyncWrite.  So, asyncFlush looks
-       //   useless.
-
-       return CompletableFuture.completedFuture(null);
-   }
-
-   @Override
-   public CompletableFuture<Void> asyncWrite(final byte[] bs, int offset, int length) {
-        final CompletableFuture<Void> cf = new CompletableFuture<>();
-        if (!started) {
-            byteSink = byteSinkSupplier.get();
-            parentResponse.committed();
-            started = true;
+public class ChunkOutputStream extends AsyncOutputStream {
+
+    /**
+     * This is the {@link Mono} that we return from {@link ReactorNettyJaxrsServer.Handler#handle(HttpServerRequest,
+     * HttpServerResponse)}
+     */
+    protected final Sinks.Empty<Void> completionSink;
+
+    /**
+     * Indicates that we've starting sending the response bytes.
+     */
+    private volatile boolean started;
+
+    private final NettyOutbound nettyOutbound;
+
+    /**
+     * Used in {@link NettyOutbound#send(Publisher, Predicate)} to trigger flushing the bytes before the {@link
+     * CompletableFuture} returned from `NettyOutbound#then().toFuture()` completes.
+     */
+    private static final Predicate<ByteBuf> FLUSH_ON_EACH_WRITE = bb -> true;
+
+    private final ReactorNettyHttpResponse parentResponse;
+
+    static final Throwable RESPONSE_WRITE_ABORTED_ON_CANCEL = new WriterException("Response write aborted due to cancellation");
+
+    static final Throwable RESPONSE_WRITE_ABORTED_ON_DISCARD = new WriterException("Response write aborted due to discarded bytes");
+
+    ChunkOutputStream(
+            final ReactorNettyHttpResponse parentResponse,
+            final HttpServerResponse reactorNettyResponse,
+            final Sinks.Empty<Void> completionSink
+    ) {
+        this.completionSink = Objects.requireNonNull(completionSink);
+        this.parentResponse = Objects.requireNonNull(parentResponse);
+        this.nettyOutbound = Objects.requireNonNull(reactorNettyResponse);
+    }
+
+    @Override
+    public void write(int b) {
+        write(new byte[]{(byte) b}, 0, 1);
+    }
+
+    @Override
+    public void write(byte[] bs, int off, int len) {
+        try {
+            asyncWrite(bs, off, len).toCompletableFuture().get();
```

Comment:
Please excuse my ignorance here as I'm not too familiar with this module :) This could potentially block forever with the `get()`. Should it have a timeout or not block at all?

Reply:
This is more or less the old code moved around, so I'll chime in on this one..  This signature comes from RestEasy's API.  We did assume it meant that the write should be completely done (at least ignoring flushing concepts) before returning.  As such, any responsibility for enforcing a timeout would be at a higher level..

Reply:
Makes sense to me. I kind of had a feeling that was it, but I figured I'd at least ask :)